File: src/main/java/com/heekng/celloct/service/JoinRequestService.java
package com.heekng.celloct.service;

import com.heekng.celloct.dto.JoinRequestDto;
import com.heekng.celloct.entity.JoinRequest;
import com.heekng.celloct.entity.Member;
import com.heekng.celloct.entity.Shop;
import com.heekng.celloct.entity.Staff;
import com.heekng.celloct.repository.JoinRequestRepository;
import com.heekng.celloct.repository.MemberRepository;
import com.heekng.celloct.repository.ShopRepository;
import com.heekng.celloct.repository.StaffRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class JoinRequestService {

    private final JoinRequestRepository joinRequestRepository;
    private final MemberRepository memberRepository;
    private final ShopRepository shopRepository;
    private final StaffRepository staffRepository;

    //가입 신청
    @Transactional
    public Long joinRequest(JoinRequestDto.joinRequest joinRequestDto) {
        Member findMember = memberRepository.findById(joinRequestDto.getMemberId()).orElseThrow(() -> new IllegalStateException("존재하지 않는 회원입니다."));
        Shop findShop = shopRepository.findById(joinRequestDto.getShopId()).orElseThrow(() -> new IllegalStateException("존재하지 않는 매장입니다."));
        validateDuplicateJoinRequest(joinRequestDto.getMemberId(), joinRequestDto.getShopId());
        validateDuplicateStaff(joinRequestDto.getMemberId(), joinRequestDto.getShopId());
        JoinRequest joinRequest = JoinRequest.builder()
                .shop(findShop)
                .member(findMember)
                .build();
        JoinRequest savedJoinRequest = joinRequestRepository.save(joinRequest);
        return savedJoinRequest.getId();
    }

    private void validateDuplicateJoinRequest(Long memberId, Long shopId) {
        List<JoinRequest> findJoinRequests = joinRequestRepository.findByMemberIdAndShopId(memberId, shopId);
        if (!findJoinRequests.isEmpty()) {
            throw new IllegalStateException("이미 가입신청한 매장입니다.");
        }
    }

    private void validateDuplicateStaff(Long memberId, Long shopId) {
        List<Staff> byMemberIdAndShopId = staffRepository.findByMemberIdAndShopId(memberId, shopId);
        if(!byMemberIdAndShopId.isEmpty()) {
            throw new IllegalStateException("이미 가입된 회원입니다.");
        }
    }

    //가입신청 조회(회원)
    public List<JoinRequest> findByMemberId(Long memberId) {
        return joinRequestRepository.findByMemberId(memberId);
    }

    //가입신청 조회(매장)
    public List<JoinRequest> findByShopId(Long shopId) {
        return joinRequestRepository.findByShopId(shopId);
    }

    //가입신청 취소
    public void cancel(Long joinRequestId) {
        JoinRequest findJoinRequest = joinRequestRepository.findById(joinRequestId).orElseThrow(() -> new IllegalStateException("존재하지 않는 가입신청입니다."));
        joinRequestRepository.delete(findJoinRequest);
    }

    //가입신청 승인
    public Long Approval(Long joinRequestId) {
        JoinRequest findJoinRequest = joinRequestRepository.findById(joinRequestId).orElseThrow(() -> new IllegalStateException("존재하지 않는 가입신청입니다."));
        Member member = findJoinRequest.getMember();
        Shop shop = findJoinRequest.getShop();
        validateDuplicateStaff(member.getId(), shop.getId());
        LocalDateTime now = LocalDateTime.now();
        Staff staff = Staff.builder()
                .shop(shop)
                .member(member)
                .build();
        staffRepository.save(staff);
        return staff.getId();
    }
}

File: src/main/kotlin/com/heekng/celloct/service/JoinRequestService.kt
package com.heekng.celloct.service

import com.heekng.celloct.dto.JoinRequestDto
import com.heekng.celloct.entity.JoinRequest
import com.heekng.celloct.entity.Staff
import com.heekng.celloct.repository.JoinRequestRepository
import com.heekng.celloct.repository.MemberRepository
import com.heekng.celloct.repository.ShopRepository
import com.heekng.celloct.repository.StaffRepository
import com.heekng.celloct.util.findByIdOrThrow
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class JoinRequestService(
    private val joinRequestRepository: JoinRequestRepository,
    private val memberRepository: MemberRepository,
    private val shopRepository: ShopRepository,
    private val staffRepository: StaffRepository,
) {

    @Transactional
    fun joinRequest(joinRequestDto: JoinRequestDto.JoinRequest): Long {
        val findMember = memberRepository.findByIdOrThrow(joinRequestDto.memberId)
        val findShop = shopRepository.findByIdOrThrow(joinRequestDto.shopId)
        validateDuplicateJoinRequest(joinRequestDto.memberId, joinRequestDto.shopId)
        validateDuplicateStaff(joinRequestDto.memberId, joinRequestDto.shopId)
        val joinRequest = JoinRequest(
            member = findMember,
            shop = findShop,
        )
        val savedJoinRequest = joinRequestRepository.save(joinRequest)
        return savedJoinRequest.id!!
    }

    @Transactional
    fun cancel(joinRequestId: Long) {
        val joinRequest = joinRequestRepository.findByIdOrThrow(joinRequestId)
        joinRequestRepository.delete(joinRequest)
    }

    @Transactional
    fun approval(approvalDto: JoinRequestDto.ApprovalRefusalRequest): Long {
        val findJoinRequest = joinRequestRepository.findByIdOrThrow(approvalDto.joinRequestId)
        val member = findJoinRequest.member
        val shop = findJoinRequest.shop
        validateDuplicateStaff(member.id!!, shop.id!!)
        val staff = Staff(
            member = member,
            shop = shop,
            name = member.name,
        )
        staffRepository.save(staff)
        joinRequestRepository.delete(findJoinRequest)
        return staff.id!!
    }

    @Transactional
    fun refusal(refusalDto: JoinRequestDto.ApprovalRefusalRequest) {
        val findJoinRequest = joinRequestRepository.findByIdOrThrow(refusalDto.joinRequestId)
        joinRequestRepository.delete(findJoinRequest)
    }

    private fun validateDuplicateJoinRequest(memberId: Long, shopId: Long) {
        val findJoinRequests = joinRequestRepository.findByMemberIdAndShopId(memberId, shopId)
        if (findJoinRequests.isNotEmpty()) {
            throw IllegalStateException("이미 가입신청한 매장입니다.")
        }
    }

    private fun validateDuplicateStaff(memberId: Long, shopId: Long) {
        val staff = staffRepository.findByMemberIdAndShopId(memberId, shopId)
        if (staff != null) {
            throw IllegalStateException("이미 가입된 회원입니다.")
        }
    }
}